File: server/tasks/hamd.py
#!/usr/bin/env python3
# hamd.py

"""
    Copyright (C) 2012-2016 Rudolf Cardinal (rudolf@pobox.com).
    Department of Psychiatry, University of Cambridge.
    Funded by the Wellcome Trust.

    This file is part of CamCOPS.

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
"""

from typing import List

from ..cc_modules.cc_db import repeat_fieldname, repeat_fieldspec
from ..cc_modules.cc_html import answer, tr, tr_qa
from ..cc_modules.cc_string import WSTRING
from ..cc_modules.cc_task import (
    CtvInfo,
    CTV_INCOMPLETE,
    get_from_dict,
    Task,
    TrackerInfo,
    TrackerLabel,
)


# =============================================================================
# HAM-D
# =============================================================================

MAX_SCORE = (
    4 * 15 -  # Q1-15 scored 0-5
    (2 * 6) +  # except Q4-6, 12-14 scored 0-2
    2 * 2  # Q16-17
)  # ... and not scored beyond Q17... total 52


class Hamd(Task):
    NSCOREDQUESTIONS = 17
    NQUESTIONS = 21

    tablename = "hamd"
    shortname = "HAM-D"
    longname = "Hamilton Rating Scale for Depression"
    fieldspecs = (
        repeat_fieldspec(
            "q", 1, 15, comment_fmt="Q{n}, {s} (scored 0-4, except 0-2 for "
            "Q4-6/12-14, higher worse)", min=0, max=4,  # amended below
            comment_strings=[
                "depressed mood", "guilt", "suicide", "early insomnia",
                "middle insomnia", "late insomnia", "work/activities",
                "psychomotor retardation", "agitation",
                "anxiety, psychological", "anxiety, somatic",
                "somatic symptoms, gastointestinal",
                "somatic symptoms, general", "genital symptoms",
                "hypochondriasis"
            ]) +
        [
            dict(name="whichq16", cctype="INT", min=0, max=1,
                 comment="Method of assessing weight loss (0 = A, by history; "
                 "1 = B, by measured change)"),
            dict(name="q16a", cctype="INT", min=0, max=3,
                 comment="Q16A, weight loss, by history (0 none - 2 definite,"
                 " or 3 not assessed [not scored])"),
            dict(name="q16b", cctype="INT", min=0, max=3,
                 comment="Q16B, weight loss, by measurement (0 none - "
                 "2 more than 2lb, or 3 not assessed [not scored])"),
            dict(name="q17", cctype="INT", min=0, max=2,
                 comment="Q17, lack of insight (0-2, higher worse)"),
            dict(name="q18a", cctype="INT", min=0, max=2,
                 comment="Q18A (not scored), diurnal variation, presence "
                 "(0 none, 1 worse AM, 2 worse PM)"),
            dict(name="q18b", cctype="INT", min=0, max=2,
                 comment="Q18B (not scored), diurnal variation, severity "
                 "(0-2, higher more severe)"),
        ] +
        repeat_fieldspec(
            "q", 19, 21, comment_fmt="Q{n} (not scored), {s} (0-4 for Q19, "
            "0-3 for Q20, 0-2 for Q21, higher worse)", min=0, max=4,  # below
            comment_strings=["depersonalization/derealization",
                             "paranoid symptoms",
                             "obsessional/compulsive symptoms"])
    )
    # Now fix the wrong bits. Hardly elegant!
    for item in fieldspecs:
        name = item["name"]
        if (name == "q4" or name == "q5" or name == "q6" or
                name == "q12" or name == "q13" or name == "q14" or
                name == "q21"):
            item["max"] = 2
        if name == "q20":
            item["max"] = 3
    has_clinician = True

    def get_trackers(self) -> List[TrackerInfo]:
        return [TrackerInfo(
            value=self.total_score(),
            plot_label="HAM-D total score",
            axis_label="Total score (out of {})".format(MAX_SCORE),
            axis_min=-0.5,
            axis_max=MAX_SCORE + 0.5,
            horizontal_lines=[22.5, 19.5, 14.5, 7.5],
            horizontal_labels=[
                TrackerLabel(25, WSTRING("hamd_severity_verysevere")),
                TrackerLabel(21, WSTRING("hamd_severity_severe")),
                TrackerLabel(17, WSTRING("hamd_severity_moderate")),
                TrackerLabel(11, WSTRING("hamd_severity_mild")),
                TrackerLabel(3.75, WSTRING("hamd_severity_none")),
            ]
        )]

    def get_clinical_text(self) -> List[CtvInfo]:
        if not self.is_complete():
            return CTV_INCOMPLETE
        return [CtvInfo(
            content="HAM-D total score {}/{} ({})".format(
                self.total_score(), MAX_SCORE, self.severity())
        )]

    def get_summaries(self):
        return [
            self.is_complete_summary_field(),
            dict(name="total", cctype="INT", value=self.total_score(),
                 comment="Total score (/{})".format(MAX_SCORE)),
            dict(name="severity", cctype="TEXT", value=self.severity(),
                 comment="Severity"),
        ]

    def is_complete(self) -> bool:
        if not self.field_contents_valid():
            return False
        if self.q1 is None or self.q9 is None or self.q10 is None:
            return False
        if self.q1 == 0:
            # Special limited-information completeness
            return True
        if self.q2 is not None and self.q3 is not None \
                and (self.q2 + self.q3 == 0):
            # Special limited-information completeness
            return True
        # Otherwise, any null values cause problems
        if self.whichq16 is None:
            return False
        for i in range(1, self.NSCOREDQUESTIONS + 1):
            if i == 16:
                if (self.whichq16 == 0 and self.q16a is None) \
                        or (self.whichq16 == 1 and self.q16b is None):
                    return False
            else:
                if getattr(self, "q" + str(i)) is None:
                    return False
        return True

    def total_score(self) -> int:
        total = 0
        for i in range(1, self.NSCOREDQUESTIONS + 1):
            if i == 16:
                relevant_field = "q16a" if self.whichq16 == 0 else "q16b"
                score = self.sum_fields([relevant_field])
                if score != 3:  # ... a value that's ignored
                    total += score
            else:
                total += self.sum_fields(["q" + str(i)])
        return total

    def severity(self) -> str:
        score = self.total_score()
        if score >= 23:
            return WSTRING("hamd_severity_verysevere")
        elif score >= 19:
            return WSTRING("hamd_severity_severe")
        elif score >= 14:
            return WSTRING("hamd_severity_moderate")
        elif score >= 8:
            return WSTRING("hamd_severity_mild")
        else:
            return WSTRING("hamd_severity_none")

    def get_task_html(self) -> str:
        score = self.total_score()
        severity = self.severity()
        task_field_list_for_display = (
            repeat_fieldname("q", 1, 15) +
            [
                "whichq16",
                "q16a" if self.whichq16 == 0 else "q16b",  # funny one
                "q17",
                "q18a",
                "q18b"
            ] +
            repeat_fieldname("q", 19, 21)
        )
        answer_dicts_dict = {}
        for q in task_field_list_for_display:
            d = {None: None}
            for option in range(0, 5):
                if (q == "q4" or q == "q5" or q == "q6" or q == "q12" or
                        q == "q13" or q == "q14" or q == "q17" or
                        q == "q18" or q == "q21") and option > 2:
                    continue
                d[option] = WSTRING("hamd_" + q + "_option" + str(option))
            answer_dicts_dict[q] = d
        h = """
            <div class="summary">
                <table class="summary">
        """ + self.get_is_complete_tr()
        h += tr(WSTRING("total_score") + " <sup>[1]</sup>",
                answer(score) + " / {}".format(MAX_SCORE))
        h += tr_qa(WSTRING("hamd_severity") + " <sup>[2]</sup>", severity)
        h += """
                </table>
            </div>
            <table class="taskdetail">
                <tr>
                    <th width="40%">Question</th>
                    <th width="60%">Answer</th>
                </tr>
        """
        for q in task_field_list_for_display:
            if q == "whichq16":
                qstr = WSTRING("hamd_whichq16_title")
            else:
                if q == "q16a" or q == "q16b":
                    rangestr = " <sup>range 0–2; ‘3’ not scored</sup>"
                else:
                    rangestr = next((
                        " <sup>range {}–{}</sup>".format(
                            item.get("min"), item.get("max")
                        )
                        for item in self.fieldspecs
                        if item["name"] == q
                    ), "")
                    # http://stackoverflow.com/questions/8653516
                qstr = WSTRING("hamd_" + q + "_s") + rangestr
            h += tr_qa(qstr, get_from_dict(answer_dicts_dict[q],
                                           getattr(self, q)))
        h += """
            </table>
            <div class="footnotes">
                [1] Only Q1–Q17 scored towards the total.
                    Re Q16: values of ‘3’ (‘not assessed’) are not actively
                    scored, after e.g. Guy W (1976) <i>ECDEU Assessment Manual
                    for Psychopharmacology, revised</i>, pp. 180–192, esp.
                    pp. 187, 189
                    (https://archive.org/stream/ecdeuassessmentm1933guyw).
                [2] ≥23 very severe, ≥19 severe, ≥14 moderate,
                    ≥8 mild, &lt;8 none.
            </div>
        """
        return h

File: tablet/Resources/task/HamD.js
// HamD.js

/*
    Copyright (C) 2012-2016 Rudolf Cardinal (rudolf@pobox.com).
    Department of Psychiatry, University of Cambridge.
    Funded by the Wellcome Trust.

    This file is part of CamCOPS.

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/

/*jslint node: true, newcap: true, nomen: true, plusplus: true */
"use strict";
/*global Titanium, L */

var DBCONSTANTS = require('common/DBCONSTANTS'),
    dbcommon = require('lib/dbcommon'),
    taskcommon = require('lib/taskcommon'),
    lang = require('lib/lang'),
    tablename = "hamd",
    fieldlist = dbcommon.standardTaskFields(),
    nscoredquestions = 17,
    qlist = [
        {name: "q1", noptions: 5, mandatory: true },
        {name: "q2", noptions: 5, mandatory: true },
        {name: "q3", noptions: 5, mandatory: true },
        {name: "q4", noptions: 3, mandatory: true },
        {name: "q5", noptions: 3, mandatory: true },
        {name: "q6", noptions: 3, mandatory: true },
        {name: "q7", noptions: 5, mandatory: true },
        {name: "q8", noptions: 5, mandatory: true },
        {name: "q9", noptions: 5, mandatory: true },
        {name: "q10", noptions: 5, mandatory: true },
        {name: "q11", noptions: 5, mandatory: true },
        {name: "q12", noptions: 3, mandatory: true },
        {name: "q13", noptions: 3, mandatory: true },
        {name: "q14", noptions: 3, mandatory: true },
        {name: "q15", noptions: 5, mandatory: true },
        {name: "whichq16", noptions: 2, mandatory: true },
        {name: "q16a", noptions: 4, mandatory: true },
        {name: "q16b", noptions: 4, mandatory: true },
        {name: "q17", noptions: 3, mandatory: true },
        {name: "q18a", noptions: 3, mandatory: false },
        {name: "q18b", noptions: 3, mandatory: false },
        {name: "q19", noptions: 5, mandatory: false },
        {name: "q20", noptions: 4, mandatory: false },
        {name: "q21", noptions: 3, mandatory: false }
    ];

fieldlist.push.apply(fieldlist, dbcommon.CLINICIAN_FIELDSPECS); // Clinician info 1/3
dbcommon.appendRepeatedFieldDef(fieldlist, "q", 1, 15,
                                DBCONSTANTS.TYPE_INTEGER);
fieldlist.push(
    {name: 'whichq16', type: DBCONSTANTS.TYPE_INTEGER},
    {name: 'q16a', type: DBCONSTANTS.TYPE_INTEGER},
    {name: 'q16b', type: DBCONSTANTS.TYPE_INTEGER},
    {name: 'q17', type: DBCONSTANTS.TYPE_INTEGER},
    {name: 'q18a', type: DBCONSTANTS.TYPE_INTEGER},
    {name: 'q18b', type: DBCONSTANTS.TYPE_INTEGER}
);
dbcommon.appendRepeatedFieldDef(fieldlist, "q", 19, 21,
                                DBCONSTANTS.TYPE_INTEGER);

// CREATE THE TABLE

dbcommon.createTable(tablename, fieldlist);

// TASK

function HamD(patient_id) {
    taskcommon.BaseTask.call(this, patient_id); // call base constructor
}

lang.inheritPrototype(HamD, taskcommon.BaseTask);
lang.extendPrototype(HamD, {

    // KEY DATABASE FIELDS (USED BY DatabaseObject)

    _objecttype: HamD,
    _tablename: tablename,
    _fieldlist: fieldlist,

    // TASK CLASS FIELD OVERRIDES (USED BY BaseTask)

    // OTHER

    // Scoring
    getTotalScore: function () {
        var total = 0,
            rawscore,
            i;
        for (i = 1; i <= nscoredquestions; ++i) {
            if (i === 16) {
                if (this.whichq16 === 0) {
                    rawscore = this.q16a;
                } else {
                    rawscore = this.q16b;
                }
                if (rawscore !== 3) {
                    // For weight questions, '3' means 'not measured' and is
                    // not scored.
                    total += rawscore;
                }
            } else {
                total += this["q" + i];
            }
        }
        return total;
    },

    // Standard task functions
    isComplete: function () {
        var i;
        if (this.q1 === null || this.q9 === null || this.q10 === null) {
            // Always need these three.
            return false;
        }
        if (this.q1 === 0) {
            // Special limited-information completeness
            return true;
        }
        if (this.q2 !== null && this.q3 !== null &&
                (this.q2 + this.q3 === 0)) {
            // Special limited-information completeness
            return true;
        }
        // Otherwise, any null values cause problems
        for (i = 1; i <= nscoredquestions; ++i) {
            if (this.whichq16 === null) {
                return false;
            }
            if (i === 16) {
                if ((this.whichq16 === 0 && this.q16a === null) ||
                        (this.whichq16 === 1 && this.q16b === null)) {
                    return false;
                }
            } else {
                if (this["q" + i] === null) {
                    return false;
                }
            }
        }
        return true;
    },

    getSummary: function () {
        return (L('hamd_total_score') + " " + this.getTotalScore() + "/52" +
                this.isCompleteSuffix());
    },

    getDetail: function () {
        var totalscore = this.getTotalScore(),
            severity = (totalscore > 23 ? L('hamd_severity_verysevere')
                        : (totalscore >= 19 ? L('hamd_severity_severe')
                           : (totalscore >= 14 ? L('hamd_severity_moderate')
                              : (totalscore >= 8 ? L('hamd_severity_mild')
                                : L('hamd_severity_none')
                                )
                             )
                          )
            ),
            msg = "",
            i;
        for (i = 0; i < qlist.length; ++i) {
            msg += (L("hamd_" + qlist[i].name + "_s") + " " +
                    this[qlist[i].name] + "\n");
        }
        return (msg +
            "\n" +
            this.getSummary() + "\n" +
            "\n" +
            L('hamd_severity') + " " + severity
        );
    },

    edit: function (readOnly) {
        var self = this,
            KeyValuePair = require('lib/KeyValuePair'),
            Questionnaire = require('questionnaire/Questionnaire'),
            pages,
            questionnaire,
            i;

        self.setDefaultClinicianVariablesAtFirstUse(readOnly); // Clinician info 2/3

        function makepage(props) {
            var options = [],
                j;
            for (j = 0; j < props.noptions; ++j) {
                options.push(new KeyValuePair(L("hamd_" + props.name +
                                                "_option" + j), j));
            }
            return {
                title: L("hamd_" + props.name + "_title"),
                clinician: true,
                elements: [
                    {
                        type: "QuestionText",
                        text: L("hamd_" + props.name + "_question")
                    },
                    {
                        type: "QuestionMCQ",
                        field: props.name,
                        mandatory: props.mandatory,
                        options: options
                    }
                ]
            };
        }
        pages = [ self.getClinicianDetailsPage() ]; // Clinician info 3/3
        for (i = 0; i < qlist.length; ++i) {
            pages.push(makepage(qlist[i]));
        }

        questionnaire = new Questionnaire({
            readOnly: readOnly,
            pages: pages,
            callbackThis: self,
            fnGetFieldValue: self.defaultGetFieldValueFn,
            fnSetField: self.defaultSetFieldFn,
            fnFinished: self.defaultFinishedFn,
            fnNextPage: function (pageId_zero_based) {
                if (pageId_zero_based === 16) { // the "whichq16" one
                    return self.whichq16 === 0 ? 17 : 18;
                }
                if (pageId_zero_based === 17 || pageId_zero_based === 18) {
                    // the weight ones
                    return 19;
                }
                return pageId_zero_based + 1;
                // by default, move to the next question
            },
            fnPreviousPage: function (pageId_zero_based) {
                if (pageId_zero_based === 17 || pageId_zero_based === 18) {
                    // one of the weight ones
                    return 16; // go to the "whichq16" one
                }
                if (pageId_zero_based === 19) {
                    // the one after the weight one
                    return self.whichq16 === 0 ? 17 : 18;
                }
                return pageId_zero_based - 1;
                // by default, move to the previous question
            }
        });

        questionnaire.open();
    }

});

module.exports = HamD;
